src/rapids_dependency_file_generator/_rapids_dependency_file_generator.py

```diff
@@ -52,18 +51,21 @@ def dedupe(
 
     Returns
     -------
-    list[str | dict[str, str]]
-        The `dependencies` with all duplicates removed.
+    list[str | dict[str, list[str]]]
+        The ``dependencies`` with all duplicates removed.
     """
-    deduped = sorted({dep for dep in dependencies if isinstance(dep, str)})
-    dict_deps = defaultdict(list)
-    for dep in filter(lambda dep: not isinstance(dep, str), dependencies):
-        if isinstance(dep, _config.PipRequirements):
-            dict_deps["pip"].extend(dep.pip)
-            dict_deps["pip"] = sorted(set(dict_deps["pip"]))
-    if dict_deps:
-        deduped.append(dict(dict_deps))
-    return deduped
+    string_deps: set[str] = set()
+    pip_deps: set[str] = set()
+    for dep in dependencies:
+        if isinstance(dep, str):
+            string_deps.add(dep)
+        elif isinstance(dep, _config.PipRequirements):
+            pip_deps.update(dep.pip)
+
+    if pip_deps:
+        return [*sorted(string_deps), {"pip": sorted(pip_deps)}]
+    else:
+        return sorted(string_deps)  # type: ignore[return-value]
```

Comment:
This `#type: ignore` is here to avoid this:

```text
src/rapids_dependency_file_generator/_rapids_dependency_file_generator.py:68: error: Incompatible return value type (got "list[str]", expected "list[str | dict[str, list[str]]]")  [return-value]
src/rapids_dependency_file_generator/_rapids_dependency_file_generator.py:68: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
src/rapids_dependency_file_generator/_rapids_dependency_file_generator.py:68: note: Consider using "Sequence" instead, which is covariant
```

`mypy` is just really unhappy with the idea of `list[str]` being a strict subset of `list[str | dict[str, list[str]]]`.

Reply:
Why not use `Sequence` instead, as the warning suggested?

Reply:
I'd tried a couple of variations using `Sequence` and then abandoned it, because doing that means you can't use methods that are common to all `Sequence` types.

e.g. I'd stuff like tried this:

```python
out: Sequence[Union[str, dict[str, list[str]]]] = sorted(string_deps)
if pip_deps:
    out.append({"pip": sorted(pip_deps)})](out: typing.Sequence[typing.Union[str, dict[str, list[str]]]] = sorted(string_deps))
```

That generated errors like this:

```text
error: "Sequence[str | dict[str, list[str]]]" has no attribute "append"
```

HOWEVER.... I never revisited that after switching to this multiple-return-statements pattern! (which avoids the `.append()`)

I just pushed https://github.com/rapidsai/dependency-file-generator/pull/90/commits/ddb8c7f53e535350369f5ba587c9b45edff23a1b switching to `Sequence` and removing this `# type: ignore`, thanks for the nudge.